rpcs3/Emu/RSX/Common/BufferUtils.cpp

```diff
@@ -755,39 +762,298 @@ namespace
+				0, 0, 0, 0, 0, 0, 0x3, 0x2);
+
+			__m128i tmp = __mm_shuffle_epi8(min, mask_step1);
+			min = __mm_min_epu16(min, tmp);
+			tmp = __mm_shuffle_epi8(min, mask_step2);
+			min = __mm_min_epu16(min, tmp);
+			tmp = __mm_shuffle_epi8(min, mask_step3);
+			min = __mm_min_epu16(min, tmp);
+
+			tmp = __mm_shuffle_epi8(max, mask_step1);
+			max = __mm_max_epu16(max, tmp);
+			tmp = __mm_shuffle_epi8(max, mask_step2);
+			max = __mm_max_epu16(max, tmp);
+			tmp = __mm_shuffle_epi8(max, mask_step3);
+			max = __mm_max_epu16(max, tmp);
+
+			const u16 min_index = u16(_mm_cvtsi128_si32(min) & 0xFFFF);
+			const u16 max_index = u16(_mm_cvtsi128_si32(max) & 0xFFFF);
+
+			return std::make_tuple(min_index, max_index, count);
+		}
+
+		SSE4_1_FUNC
+		static
+		std::tuple<u32, u32, u32> upload_u32_swapped_sse4_1(const void *src, void *dst, u32 count, u32 restart_index)
+		{
+			u32 dst_index = 0;
+
+			const __m128i mask = _mm_set_epi8(
+				0xC, 0xD, 0xE, 0xF,
+				0x8, 0x9, 0xA, 0xB,
+				0x4, 0x5, 0x6, 0x7,
+				0x0, 0x1, 0x2, 0x3);
+
+			auto src_stream = (const __m128i*)src;
+			auto dst_stream = (__m128i*)dst;
+
+			__m128i restart = _mm_set1_epi32(restart_index);
+			__m128i min = _mm_set1_epi32(0xffffffff);
+			__m128i max = _mm_set1_epi32(0);
+
+			const auto iterations = count / 4;
+			for (unsigned n = 0; n < iterations; ++n)
+			{
+				const __m128i raw = _mm_loadu_si128(src_stream++);
+				const __m128i value = __mm_shuffle_epi8(raw, mask);
+				const __m128i mask = _mm_cmpeq_epi32(restart, value);
+				const __m128i tmp = _mm_andnot_si128(mask, value);
+				max = __mm_max_epu32(max, tmp);
+				min = __mm_min_epu32(min, value);
+				_mm_storeu_si128(dst_stream++, value);
+			}
+
+			const __m128i mask_step1 = _mm_set_epi8(
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0xF, 0xE, 0xD, 0xC, 0xB, 0xA, 0x9, 0x8);
+
+			const __m128i mask_step2 = _mm_set_epi8(
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0, 0, 0, 0, 0x7, 0x6, 0x5, 0x4);
+
+			__m128i tmp = __mm_shuffle_epi8(min, mask_step1);
+			min = __mm_min_epu32(min, tmp);
+			tmp = __mm_shuffle_epi8(min, mask_step2);
+			min = __mm_min_epu32(min, tmp);
+
+			tmp = __mm_shuffle_epi8(max, mask_step1);
+			max = __mm_max_epu32(max, tmp);
+			tmp = __mm_shuffle_epi8(max, mask_step2);
+			max = __mm_max_epu32(max, tmp);
+
+			const u32 min_index = u32(_mm_cvtsi128_si32(min));
+			const u32 max_index = u32(_mm_cvtsi128_si32(max));
+
+			return std::make_tuple(min_index, max_index, count);
+		}
+
+		template<typename T>
+		static
+		std::tuple<T, T, u32> upload_untouched(gsl::span<to_be_t<const T>> src, gsl::span<T> dst, T restart_index, bool skip_restart)
+		{
+			T min_index, max_index;
+			u32 written;
+			u32 remaining = src.size();
+
+			if (s_use_avx2 && remaining >= 32 && !skip_restart)
 			{
+				if constexpr (std::is_same<T, u32>::value)
+				{
+					const auto count = (remaining & ~0x3);
+					std::tie(min_index, max_index, written) = upload_u32_swapped_sse4_1(src.data(), dst.data(), count, restart_index);
```

Comment:
Why is this sse4.1 method guarded by a use_avx2 guard? I know there is another block before but this is essentially dead code.

```
if (remaining >= 32 && !skip_restart)
{
    if constexpr (is_same<T, u16>::value)
    {
        if (avx2) ..call avx2 path
        if (sse41) call sse4.1 path 
    }
    else
    {
        if (sse4.1) call sse4.1 path for u32
    }
}
...
fallback
```

Reply:
Sensible, I’ll fix that shortly.

Reply:
This also fixes the hypothetical case of a CPU with AVX2 support but no SSE4.1, which can probably be created using qemu.